generate/generate_report.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package generate
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/mineiros-io/terramate/stack"
+)
+
+// Result represents code generation result
+type Result struct {
+	// StackPath is the absolute path of the stack relative to the project root.
+	StackPath string
+	// Created contains filenames of all created files inside the stack
+	Created []string
+	// Changed contains filenames of all changed files inside the stack
+	Changed []string
+	// Deleted contains filenames of all deleted files inside the stack
+	Deleted []string
+}
+
+// FailureResult represents a failure on code generation.
+type FailureResult struct {
+	Result
+	Error error
+}
+
+// Report has the results of the code generation process.
+type Report struct {
+	// BootstrapErr is an error that happened before code generation
+	// could be started, indicating that no changes were made to any stack.
+	BootstrapErr error
+
+	// Successes are the success results
+	Successes []Result
+
+	// Failures are stacks that failed without generating any code
+	Failures []FailureResult
+}
+
+func (r Report) HasFailures() bool {
+	return r.BootstrapErr != nil || len(r.Failures) > 0
+}
+
+func (r Report) String() string {
+	if r.empty() {
+		return "Nothing to do, code generation is updated"
```

Comment:
As tests rely on this message, why not put it in an exported const?

Reply:
The whole formatting of the report is also not exported properly (maybe if we used a template), so it seemed OK, if you change the formatting of the report only the formatting related tests will fail etc (which are not a lot). The main rationale being that I usually don't like exporting things that I dont see any usage beyond testing, like adding things to an API for test purposes only, and I could not imagine why someone would want to check programatically that the string representation is for an empty report. Like you have better ways to check an empty report, in this case exporting the empty method would seem better, then code can just call report.Empty().